proc/process.c

```diff
@@ -877,93 +875,76 @@ int process_load(process_t *process, vm_object_t *o, off_t base, size_t size, vo
 	/* find symtab */
 	for (i = 0, shdr = (void *)((char *)ehdr + ehdr->e_shoff); i < ehdr->e_shnum; i++, shdr++) {
 		if (hal_strcmp(&snameTab[shdr->sh_name], ".symtab") == 0) {
+			symTab = (void *)((ptr_t)ehdr + (ptr_t)shdr->sh_offset);
 			break;
 		}
 	}
 
-	if (i >= ehdr->e_shnum) {
-		return -ENOEXEC;
-	}
-	symTab = (ptr_t)ehdr + (ptr_t)shdr->sh_offset;
-
 	/* Perform data, init_array and fini_array relocation */
 	for (i = 0, shdr = (void *)((char *)ehdr + ehdr->e_shoff); i < ehdr->e_shnum; i++, shdr++) {
-		/* strncmp as there may be multiple .rela.* sections for different sections.  */
-		if (hal_strncmp(&snameTab[shdr->sh_name], ".rela", 5) != 0) {
-			continue;
-		}
-
 		if ((shdr->sh_size == 0) || (shdr->sh_entsize == 0)) {
 			continue;
 		}
 
-		for (j = 0; j < shdr->sh_size / shdr->sh_entsize; ++j) {
-			rela = (Elf32_Rela *)((ptr_t)shdr->sh_offset + (ptr_t)ehdr + (j * shdr->sh_entsize));
-			reltype = ELF32_R_TYPE(rela->r_info);
-
-			if (reltype == R_SPARC_32) {
-				relptr = (void *)rela->r_offset;
-				if (process_relocate(reloc, relocsz, (char **)&relptr) < 0) {
-					return -ENOEXEC;
-				}
-
-				/* Don't modify ELF file! */
-				if (((ptr_t)relptr >= (ptr_t)base) && ((ptr_t)relptr < ((ptr_t)base + size))) {
-					++badreloc;
-					continue;
-				}
-
-				sym = (Elf32_Sym *)(symTab + (ELF32_R_SYM(rela->r_info) * sizeof(Elf32_Sym)));
-
-				/* Write addend to the address */
-				*(char **)relptr = (char *)(sym->st_value + rela->r_addend);
-
-				if (process_relocate(reloc, relocsz, relptr) < 0) {
-					return -ENOEXEC;
-				}
-			}
+		/* strncmp as there may be multiple .rela.* or .rel.* sections for different sections.  */
+		if (hal_strncmp(&snameTab[shdr->sh_name], ".rela.", 6) == 0) {
+			isrela = 1;
 		}
-	}
-#else
-	/* Perform data, init_array and fini_array relocation */
-	for (i = 0, shdr = (void *)((char *)ehdr + ehdr->e_shoff); i < ehdr->e_shnum; i++, shdr++) {
-		/* strncmp as there may be multiple .rel.* sections for different sections. */
-		if (hal_strncmp(&snameTab[shdr->sh_name], ".rel", 4) != 0) {
-			continue;
+		else if (hal_strncmp(&snameTab[shdr->sh_name], ".rel.", 5) == 0) {
+			isrela = 0;
 		}
-
-		if ((shdr->sh_size == 0) || (shdr->sh_entsize == 0)) {
+		else {
 			continue;
 		}
 
-		for (j = 0; j < shdr->sh_size / shdr->sh_entsize; ++j) {
-			rel = (void *)((ptr_t)shdr->sh_offset + (ptr_t)ehdr + (j * shdr->sh_entsize));
-			reltype = ELF32_R_TYPE(rel->r_info);
+		for (j = 0; j < (shdr->sh_size / shdr->sh_entsize); ++j) {
+			/* Valid for both Elf32_Rela and Elf32_Rel, due to correct size being stored in shdr->sh_entsize. */
+			/* For .rel. section make sure not to access addend field! */
+			hal_memcpy(&rela, (Elf32_Rela *)((ptr_t)shdr->sh_offset + (ptr_t)ehdr + (j * shdr->sh_entsize)), shdr->sh_entsize);
+			reltype = ELF32_R_TYPE(rela.r_info);
 
-			if (reltype == R_ARM_ABS32 || reltype == R_ARM_TARGET1) {
-				relptr = (void *)rel->r_offset;
+			if (hal_isRelReloc(reltype) == 0) {
```

Comment:
`reltype` became redundant in the current approach
```suggestion
			if (hal_isRelReloc(ELF32_R_TYPE(rela.r_info)) == 0) {
```